File: Fatec_DSM_3/TecProg2/Aulas/Aula_09_08_24/PadroesCriacionais/FactoryMethod/FactoryMethod01.py
# Factory Method em Python

# Classe Veículo
class Veiculo:

# Construtor __init__

    # O "def" cria uma função
    def __init__(self, modelo):
        self.modelo = modelo # O "self" é o mesmo que "this"

    def mostrarDetalhes(self):
        print(f'Modelo: {self.modelo}')    

# Subclasses de Veículo

# A classe "Carro" está herdando o que está em "Veiculo"
class Carro(Veiculo):
    def __init__(self, modelo):
        # O "super" está chamando a classe "Veiculo"
        super().__init__(modelo)

class Moto(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

class Barco(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

# Fábrica Abstrata de Veículos
class FabricaDeVeiculos:
    def criarVeiculo(self, modelo):
        # O "raise" é uma exceção de erro
        raise NotImplementedError('O método criarVeiculo() deve ser implementado pelas subclasses')
    
# Fábrica Concreta de Carros
class FabricaDeCarros(FabricaDeVeiculos):
    def criarVeiculo(self, modelo):
        return Carro(modelo)

# Fábrica Concreta de Motos
class FabricaDeMotos(FabricaDeVeiculos):
    def criarVeiculo(self, modelo):
        return Moto(modelo)

# Fábrica Concreta de Barcos
class FabricaDeBarcos(FabricaDeVeiculos):
    def criarVeiculo(self, modelo):
        return Barco(modelo)    
    
# Uso do Factory Method

fabricaCarros = FabricaDeCarros()
carro1 = fabricaCarros.criarVeiculo('Sedan')
carro2 = fabricaCarros.criarVeiculo('Fusca 68')
carro1.mostrarDetalhes() # Saída: "Modelo: Sedan"
carro2.mostrarDetalhes() # Saída: "Modelo: Fusca 68"

fabricaMotos = FabricaDeMotos()
moto1 = fabricaMotos.criarVeiculo('Esportiva')
moto1.mostrarDetalhes() # Saída: "Modelo: Esportiva"

fabricaBarcos = FabricaDeBarcos()
barco1 = fabricaBarcos.criarVeiculo('Lancha')
barco1.mostrarDetalhes() # Saída: "Modelo: Lancha"

File: Fatec_DSM_3/TecProg2/Aulas/Aula_09_08_24/PadroesCriacionais/FactoryMethod/FactoryMethod01.java
// Factory Method em Java

// Classe Veículo
class Veiculo{

    private String modelo;

    public Veiculo(String modelo){
        this.modelo = modelo;
    }

    public void mostrarDetalhes(){
        System.out.println("Modelo: " + this.modelo);
    }
}

// Subclasses de Veículo

class Carro extends Veiculo{

    public Carro(String modelo){
        super(modelo);
    }
}

class Moto extends Veiculo{

    public Moto(String modelo){
        super(modelo);
    }
}

class Barco extends Veiculo{

    public Barco(String modelo){
        super(modelo);
    }
}

// Fábrica Abstrata de Veículos
abstract class FabricaDeVeiculos{

    public abstract Veiculo criarVeiculo(String modelo);
}

// Fábrica Concreta de Carros
class FabricaDeCarros extends FabricaDeVeiculos{

    public Veiculo criarVeiculo(String modelo){
        return new Carro(modelo);
    }
}

// Fábrica Concreta de Motos
class FabricaDeMotos extends FabricaDeVeiculos{

    public Veiculo criarVeiculo(String modelo){
        return new Moto(modelo);
    }
}

// Fábrica Concreta de Barcos
class FabricaDeBarcos extends FabricaDeVeiculos{

    public Veiculo criarVeiculo(String modelo){
        return new Barco(modelo);
    }
}

// Utilização do Factory Method

public class FactoryMethod01{

    public static void main(String[] args) {
        
        FabricaDeVeiculos fabricaCarros = new FabricaDeCarros();
        Veiculo carro1 = fabricaCarros.criarVeiculo("Sedan");
        Veiculo carro2 = fabricaCarros.criarVeiculo("Fusca 68");
        carro1.mostrarDetalhes(); // Saída: "Modelo: Sedan"
        carro2.mostrarDetalhes(); // Saída: "Modelo: Fusca 68"

        FabricaDeVeiculos fabricaMotos = new FabricaDeMotos();
        Veiculo moto1 = fabricaMotos.criarVeiculo("Esportiva");
        moto1.mostrarDetalhes(); // Saída: "Modelo: Esportiva"

        FabricaDeBarcos fabricaBarcos = new FabricaDeBarcos();
        Veiculo barco1 = fabricaBarcos.criarVeiculo("Lancha");
        barco1.mostrarDetalhes(); // Saída: "Modelo: Lancha"
    }
}
